src/main/java/gregtech/api/capability/DualHandler.java

```diff
@@ -0,0 +1,277 @@
+package gregtech.api.capability;
+
+import gregtech.api.capability.impl.FluidTankList;
+import gregtech.api.metatileentity.MetaTileEntity;
+
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidTankInfo;
+import net.minecraftforge.fluids.IFluidTank;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+import net.minecraftforge.items.IItemHandlerModifiable;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DualHandler implements IItemHandlerModifiable, IMultipleTankHandler, INotifiableHandler {
+
+    @NotNull
+    IItemHandlerModifiable itemDelegate;
+    @NotNull
+    IMultipleTankHandler fluidDelegate;
+
+    List<MetaTileEntity> notifiableEntities = new ArrayList<>();
+    private final boolean isExport;
+
+    public DualHandler(@NotNull IItemHandlerModifiable itemDelegate,
+                       @NotNull IMultipleTankHandler fluidDelegate,
+                       boolean isExport) {
+        this.itemDelegate = itemDelegate;
+        this.fluidDelegate = fluidDelegate;
+        this.isExport = isExport;
+    }
+
+    public DualHandler(@NotNull IItemHandlerModifiable itemDelegate,
+                       @NotNull IFluidTank fluidTank,
+                       boolean isExport) {
+        this.itemDelegate = itemDelegate;
+        this.fluidDelegate = new FluidTankList(false, fluidTank);
+        this.isExport = isExport;
+    }
+
+    public IItemHandlerModifiable getItemDelegate() {
+        return this.itemDelegate;
+    }
+
+    public IMultipleTankHandler getFluidDelegate() {
+        return this.fluidDelegate;
+    }
+
+    public boolean isExport() {
+        return this.isExport;
+    }
+
+    @Override
+    public int getSlots() {
+        return itemDelegate.getSlots();
+    }
+
+    @Override
+    public ItemStack getStackInSlot(int slot) {
+        return itemDelegate.getStackInSlot(slot);
+    }
+
+    @Override
+    public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
+        if (!simulate) onContentsChanged();
```

Comment:
Should `onContentsChanged` be called after the actual operation instead of before? There's a bunch of places like this in this class.